one/src/ethereum_rpc.rs

```diff
@@ -0,0 +1,175 @@
+        Self {
+            url,
+            client: reqwest::Client::new(),
+        }
+    }
+
+    async fn eth_block_number(&self) -> Result<serde_json::Value> {
+        // Get the latest block number.
+        // curl https://mainnet.infura.io/v3/{api_token} \
+        // -X POST \
+        // -H "Content-Type: application/json" \
+        // -d '{"jsonrpc":"2.0","method":"eth_blockNumber","params": [],"id":1}'
+        // >> {"jsonrpc": "2.0", "id": 1, "result": "0x127cc18"}
+        Ok(self
+            .client
+            .post(&self.url)
+            .json(&serde_json::json!({
+                "jsonrpc": "2.0",
+                "method": "eth_blockNumber",
+                "params": [],
+                "id": 1,
+            }))
+            .send()
+            .await?
+            .json()
+            .await?)
+    }
+
+    async fn eth_block_by_hash(&self, block_hash: &str) -> Result<serde_json::Value> {
+        // curl https://mainnet.infura.io/v3/{api_token} \
+        //     -X POST \
+        //     -H "Content-Type: application/json" \
+        //     -d '{"jsonrpc":"2.0","method":"eth_getBlockByHash","params": ["0x{block_hash}",false],"id":1}'
+        // >> {"jsonrpc": "2.0", "id": 1, "result": {"number": "0x105f34f", "timestamp": "0x644fe98b"}}
+        Ok(self
+            .client
+            .post(&self.url)
+            .json(&serde_json::json!({
+                "jsonrpc": "2.0",
+                "method": "eth_getBlockByHash",
+                "params": [block_hash, false],
+                "id": 1,
+            }))
+            .send()
+            .await?
+            .json()
+            .await?)
+    }
+
+    async fn eth_transaction_by_hash(&self, transaction_hash: &str) -> Result<serde_json::Value> {
+        // Get the block_hash and input from the transaction.
+        // curl https://mainnet.infura.io/v3/{api_token} \
+        //   -X POST \
+        //   -H "Content-Type: application/json" \
+        //   -d '{"jsonrpc":"2.0","method":"eth_getTransactionByHash",
+        //        "params":["0xBF7BC715A09DEA3177866AC4FC294AC9800EE2B49E09C55F56078579BFBBF158"],"id":1}'
+        // >> {"jsonrpc":"2.0", "id":1, "result": {
+        //       "blockHash": "0x783cd5a6febe13d08ac0d59fa7e666483d5e476542b29688a6f0bec3d15febd4",
+        //       "blockNumber": "0x105f34f",
+        //       "input": "0x97ad09eb41b6408c1b4be5016f652396ef47c0982c36d5877ebb874919bae3a9b854d8e1"
+        //    }}
+        Ok(self
+            .client
+            .post(&self.url)
+            .json(&serde_json::json!({
+                "jsonrpc": "2.0",
+                "method": "eth_getTransactionByHash",
+                "params": [transaction_hash],
+                "id": 1,
+            }))
+            .send()
+            .await?
+            .json()
+            .await?)
+    }
+}
+
+impl EthRpc for HttpEthRpc {
+    async fn root_time_by_transaction_cid(&self, cid: Cid) -> Result<RootTime> {
```

Comment:
I don't usually see ethereum transaction ids encoded as CIDs. Any reason we're using a Cid for this and then converting to tx_hash instead of just passing around the tx_hash everywhere instead?

Reply:
We use CIDs to express the roots because that is what is in the 'proof' field of the Time Event. The problem is that this function needs a EthRpc to get the data from eth network but the fact that it is the Timestamp and the root that we want is ceramic logic not EthRPC logic so it should be moved to a function in Events that takes an EthRPC.